node/bindnode/node.go

```diff
@@ -0,0 +1,1103 @@
+			AppropriateKind: ipld.KindSet{ipld.Kind_String},
+			ActualKind:      w.kind(),
+		}
+	}
+	w.nonPtrVal().SetString(s)
+	if w.finish != nil {
+		if err := w.finish(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *_assembler) AssignBytes(p []byte) error {
+	if w.kind() != ipld.Kind_Bytes {
+		return ipld.ErrWrongKind{
+			TypeName:        w.schemaType.Name().String(),
+			MethodName:      "AssignBytes",
+			AppropriateKind: ipld.KindSet{ipld.Kind_Bytes},
+			ActualKind:      w.kind(),
+		}
+	}
+	w.nonPtrVal().SetBytes(p)
+	return nil
+}
+
+func (w *_assembler) AssignLink(link ipld.Link) error {
+	newVal := reflect.ValueOf(link)
+	if !newVal.Type().AssignableTo(w.val.Type()) {
+		return ipld.ErrWrongKind{
+			TypeName:   w.schemaType.Name().String(),
+			MethodName: "AssignLink",
+			// TODO
+		}
+	}
+	w.nonPtrVal().Set(newVal)
+	if w.finish != nil {
+		if err := w.finish(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *_assembler) AssignNode(node ipld.Node) error {
+	// TODO: does this ever trigger?
+	// newVal := reflect.ValueOf(node)
+	// if newVal.Type().AssignableTo(w.val.Type()) {
+	// 	w.val.Set(newVal)
+	// 	return nil
+	// }
+	switch node.Kind() {
+	case ipld.Kind_Map:
+		itr := node.MapIterator()
+		am, err := w.BeginMap(-1) // TODO: length?
+		if err != nil {
+			return err
+		}
+		for !itr.Done() {
+			k, v, err := itr.Next()
+			if err != nil {
+				return err
+			}
+			if err := am.AssembleKey().AssignNode(k); err != nil {
+				return err
+			}
+			if err := am.AssembleValue().AssignNode(v); err != nil {
+				return err
+			}
+		}
+		return am.Finish()
+	case ipld.Kind_List:
+		itr := node.ListIterator()
+		am, err := w.BeginList(-1) // TODO: length?
+		if err != nil {
+			return err
+		}
+		for !itr.Done() {
+			_, v, err := itr.Next()
+			if err != nil {
+				return err
+			}
+			if err := am.AssembleValue().AssignNode(v); err != nil {
+				return err
+			}
+		}
+		return am.Finish()
+
+	case ipld.Kind_Int:
+		i, err := node.AsInt()
+		if err != nil {
+			return err
+		}
+		return w.AssignInt(i)
+	case ipld.Kind_String:
+		s, err := node.AsString()
+		if err != nil {
+			return err
+		}
+		return w.AssignString(s)
+	case ipld.Kind_Bytes:
+		p, err := node.AsBytes()
+		if err != nil {
+			return err
+		}
+		return w.AssignBytes(p)
+	case ipld.Kind_Link:
+		l, err := node.AsLink()
+		if err != nil {
+			return err
+		}
+		return w.AssignLink(l)
+	case ipld.Kind_Null:
+		return w.AssignNull()
+	}
+	// fmt.Println(w.val.Type(), reflect.TypeOf(node))
+	panic(fmt.Sprintf("TODO: %v %v", w.val.Type(), node.Kind()))
+}
+
+func (w *_assembler) Prototype() ipld.NodePrototype {
+	panic("TODO: Assembler.Prototype")
+}
+
+type _structAssembler struct {
+	// TODO: embed _assembler?
+
+	schemaType *schema.TypeStruct
+	val        reflect.Value // non-pointer
+	finish     func() error
+
+	// TODO: more state checks
+	doneFields []bool // TODO: replace with just reflect.Value field pointer checks
```

Comment:
I think this TODO is probably overoptimistic.  If a field has either the nullable or optional properties in the schema, checking the go field for nil doesn't provide any useful information about if it was assigned or not.

Keeping a bitfield explicitly is also just going to be a lot simpler code, I suspect.

Reply:
Ah, you're right, I keep forgetting about AssignNull which does assign into a node, but explicitly keeps it null.

I still think there's room for improvement here potentially, but I also don't think it's going to be a bottleneck soon. It would always be possible to avoid an extra allocation for most structs by backing this slice with a `[16]bool` in the struct., for example.

I'll leave this TODO about reminding myself to look at this at some point.